support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala

```diff
@@ -1,78 +1,242 @@
+    payment_amount: String,
+    country: String,
+    date_of_first_payment: String,
+    currency: String,
+    trial_period: String,
+    paymentFieldsAttributes: PaymentFieldsAttributes,
+  ) extends DigitalSubscriptionEmailAttributes
+
+  case class GifteeRedemptionAttributes(
+    gift_recipient_first_name: String,
+    subscription_details: String,
+    gift_start_date: String,
+    gift_recipient_email: String,
+  ) extends DigitalSubscriptionEmailAttributes
+
+  case class GifterPurchaseAttributes(
+    gifter_first_name: String,
+    gifter_last_name: String,
+    gift_recipient_first_name: String,
+    gift_recipient_last_name: String,
+    gift_recipient_email: String,
+    gift_personal_message: String,
+    gift_code: String,
+    gift_delivery_date: String,
+    subscription_details: String,
+    paymentAttributes: PaymentFieldsAttributes,
+    date_of_first_payment: String,
+  ) extends DigitalSubscriptionEmailAttributes
+
+  case class GifteeNotificationAttributes(
+    gifter_first_name: String,
+    gift_personal_message: String,
+    gift_code: String,
+  ) extends DigitalSubscriptionEmailAttributes
+
+}
+class DigitalPackEmailFields(
+  subscriptionEmailFields: SubscriptionEmailFields,
+) {
+
+  import DigitalPackEmailFields._
+  import DigitalSubscriptionEmailAttributes._
+  import subscriptionEmailFields._
+  import allProductsEmailFields._
+
+  private def directOrCorpFields(details: String) = BasicDSAttributes(
+    zuorasubscriberid = subscriptionNumber,
+    emailaddress = user.primaryEmailAddress,
+    first_name = user.firstName,
+    last_name = user.lastName,
+    subscription_details = details
+  )
 
   def build(
-    subscriptionEmailFields: SubscriptionEmailFields,
-    paidSubPaymentData: Option[PaymentMethodWithSchedule]
-  ): EmailFields = {
-    import subscriptionEmailFields._
-    import allProductsEmailFields._
-
-    val fieldsForReaderType =
-      paidSubPaymentData match {
-        case Some(PaymentMethodWithSchedule(pm, paymentSchedule)) =>
-          paymentFields(pm, directDebitMandateId) ++
-            directReaderFields(promotion, billingPeriod, user, currency, paymentSchedule)
-        case None /*Corporate*/ => List(
-          "subscription_details" -> "Group subscription"
-        )
-      }
-
-    val fields = List(
-      "zuorasubscriberid" -> subscriptionNumber,
-      "emailaddress" -> user.primaryEmailAddress,
-      "first_name" -> user.firstName,
-      "last_name" -> user.lastName,
-    ) ++ fieldsForReaderType
-
-    val dataExtensionName = if (paidSubPaymentData.isDefined) "digipack" else "digipack-corp"
-
-    EmailFields(fields, Left(sfContactId), user.primaryEmailAddress, dataExtensionName)
+    paidSubPaymentData: Option[PaymentMethodWithSchedule],
+    readerType: ReaderType
+  ): Either[String, List[EmailFields]] =
+    (paidSubPaymentData, readerType) match {
+      case (Some(PaymentMethodWithSchedule(pm, paymentSchedule)), ReaderType.Gift) =>
```

Comment:
Is there some way to label what each of these cases means? Maybe break out the result into a separate function?
```scala
case (Some(PaymentMethodWithSchedule(pm, paymentSchedule)), ReaderType.Gift) => giftPurchase(pm)
case (Some(PaymentMethodWithSchedule(pm, paymentSchedule)), _) => purchase(pm, paymentSchedule)
```
Not sure if that will just add clutter...

Reply:
yeah actually it's the unclear use of Some/None that's making it hard for me.

another option is to not use "Option" and roll a clearly named type (or even a type alias if that works?), so we would have 
```scala
case (Purchase(purchaseInfo), ReaderType.Gift) => giftPurchase(purchaseInfo.pm)
case (Purchase(purchaseInfo), _) => purchase(purchaseInfo)
case (Redemption, ReaderType.Corporate) => wrap("digipack-corp", buildCorpRedemptionAttributes).map(List(_))
case (Redemption, ReaderType.Gift) => wrap("digipack-gift-redemption", buildGiftRedemptionAttributes).map(List(_))
case (Redemption, _) => Left("can only redeem gift and corporate subs")
```

Reply:
I'll first try moving the call to "wrap" into the functions that already exist, see how that helps.

Reply:
turns out it's trivial to alias, seems to work locally
```scala
  def build(
    paidSubPaymentData: Option[PaymentMethodWithSchedule],
    readerType: ReaderType
  ): Either[String, List[EmailFields]] = {

    val Purchase = Some
    val Redemption = None

    (paidSubPaymentData, readerType) match {
      case (Purchase(paymentInfo), ReaderType.Gift) =>
        List(
          buildGiftPurchaseAttributes(paymentInfo.paymentMethod),
          buildGiftNotificationAttributes
        ).sequence
      case (Purchase(paymentInfo), _) => directAttributes(paymentInfo).map(List(_))
      case (Redemption, ReaderType.Corporate) => buildCorpRedemptionAttributes.map(List(_))
      case (Redemption, ReaderType.Gift) => buildGiftRedemptionAttributes.map(List(_))
      case (Redemption, _) => Left("can only redeem gift and corporate subs")
    }
  }
```